### PR TITLE
porting diag_second_moments_lbycond function

### PR DESCRIPTION
**Pull request by xyuan**
porting diag_second_moments_lbycond function

BFB

### CODE REVIEW

File: components/cam/src/physics/cam/shoc.F90 [hunk 1]
```diff
@@ -1154,7 +1167,7 @@ subroutine diag_second_moments_lbycond(&
     wqw_sec(i) = wqw_sfc(i)
     uw_sec(i) = uw_sfc(i)
     vw_sec(i) = vw_sfc(i)
-    wtke_sec(i) = max(sqrt(ustar2(i)),0.01_rtype)**3
+    wtke_sec(i) = bfb_pow(max(bfb_sqrt(ustar2(i)),0.01_rtype), 3.0_rtype)
```
Comment (jgfouca) [inline components/cam/src/physics/cam/shoc.F90:1170]: You should use bfb_cube here

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+KOKKOS_FUNCTION
+void Functions<S,D>
+::shoc_diag_second_moments_lbycond(
+    const Scalar& wthl_sfc, const Scalar& wqw_sfc, const Scalar& uw_sfc, const Scalar& vw_sfc,
+    const Scalar& ustar2, const Scalar& wstar,
+    Scalar& wthl_sec, Scalar& wqw_sec, Scalar& uw_sec, Scalar& vw_sec,
+    Scalar& wtke_sec, Scalar& thl_sec, Scalar& qw_sec, Scalar& qwthl_sec)
+{
+
+//printf("c++-1 %16.9f, %16.9f, %16.9f, %16.9f, %16.9f, %16.9f, %16.9f\n",wthl_sfc(0)[0],wqw_sfc(0)[0],ustar2(0)[0],wstar(0)[0],thl_sec(0)[0],uw_sfc(0)[0],vw_sfc(0)[0]);
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:21]: Remove this commented-out printf.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  qw_sec    = zero_p_four*one_p_eight*pow(wqw_sfc/uf, two); 
+  qwthl_sec = zero_p_two*one_p_eight*(wthl_sfc/uf)*(wqw_sfc/uf);
+
+  // Vertical fluxes of heat and moisture, simply
+  // use the surface fluxes given by host model
+  wthl_sec = wthl_sfc;
+  wqw_sec  = wqw_sfc;
+  uw_sec   = uw_sfc;
+  vw_sec   = vw_sfc;
+  wtke_sec = pow(max(sqrt(ustar2),ufmin), three);
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:54]: Use cube.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  const auto one_p_eight = 1.8;
+  const auto ufmin = 0.01;
+  const auto zero_p_two   = 0.2;
+  const auto zero_p_three = 0.3;
+  const auto zero_p_four  = 0.4;
+  const auto two   = 2.0;
+  const auto three = 3.0;
+
+
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:39]: Remove extra newline

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  // Purpose of this subroutine is to diagnose the lower
+  // boundary condition for the second order moments needed
+  // for the SHOC parameterization.
+  // The thermodymnamic, tracer, and momentum fluxes are set
+  // to the surface fluxes for the host model, while the
+  // thermodynamic variances and covariances are computed
+  // according to that of Andre et al. 1978.
+
+  const auto one_p_eight = 1.8;
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:31]: I'm not sure I like this approach of having every literal set to a constant. I would just use the literals inline like is done in fortran. In any case, these cannot be `const auto` because C++ will assume type double for all of these, which will break bfb in single precision. They need to be `const Scalar` if you're going to do things this way. If you're going to go back to inline literals, which I would prefer, you'll need to use the scream-precsion macro, IE `sp(1.8)`

File: components/cam/src/physics/cam/shoc.F90 [hunk 6]
```diff
@@ -1135,16 +1135,29 @@ subroutine diag_second_moments_lbycond(&
...
   do i=1,shcol
 
-    uf = sqrt(ustar2(i) + 0.3_rtype * wstar(i) * wstar(i))
+    uf = bfb_sqrt(ustar2(i) + 0.3_rtype * wstar(i) * wstar(i))
     uf = max(ufmin,uf)
 
     ! Diagnose thermodynamics variances and covariances
-    thl_sec(i) = 0.4_rtype * a_const * (wthl_sfc(i)/uf)**2
-    qw_sec(i) = 0.4_rtype * a_const * (wqw_sfc(i)/uf)**2
+    thl_sec(i) = 0.4_rtype * a_const * bfb_pow(wthl_sfc(i)/uf, 2.0_rtype)
```
Comment (jgfouca) [inline components/cam/src/physics/cam/shoc.F90:1159]: Use bfb_square.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  const auto zero_p_four  = 0.4;
+  const auto two   = 2.0;
+  const auto three = 3.0;
+
+
+  auto uf = sqrt(ustar2+zero_p_three*wstar*wstar);
+  uf = max(ufmin,uf);
+
+  // Diagnose thermodynamics variances and covariances
+  thl_sec   = zero_p_four*one_p_eight*pow(wthl_sfc/uf, two);
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:44]: Use square.

File: components/scream/src/physics/shoc/tests/shoc_diag_second_mom_lbycond_test.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  SHOCSecondMomLbycondData lby_fortran[] = {
+    // shcol
+    SHOCSecondMomLbycondData(120),
+    SHOCSecondMomLbycondData(120),
+    SHOCSecondMomLbycondData(120),
+    SHOCSecondMomLbycondData(120),
+  };
+
+  static constexpr Int num_runs = sizeof(lby_fortran) / sizeof(SHOCSecondMomLbycondData);
+
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_diag_second_mom_lbycond_test.cpp:37]: Need to generate some random input data here (randomize method on the Data objects), otherwise we are just testing against all-zero input, which is why the bfb problems were not caught.

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp [hunk 9]
```diff
@@ -407,6 +410,15 @@ void linear_interp(SHOCLinearintData& d)
   d.transpose<ekat::util::TransposeDirection::f2c>();
 }
 
+void shoc_diag_second_moments_lbycond(SHOCSecondMomLbycondData& d)
+{
+  shoc_init(d.nlev, true);
+  d.transpose<ekat::util::TransposeDirection::c2f>();
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_functions_f90.cpp:416]: Since the data is 1-dimensional, we don't need to transpose anything, although it's harmless to make the call.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp (hunk 5 above)
Comment (jeff-cohere) [reply components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:31]: I also think using literals is better for clarity and less prone to unintended consequences related to `auto`. `sp(x)` isn't great, but it's not hard to understand, and I for one prefer the formulae to have numerical coefficients visible.

File: components/scream/src/physics/shoc/shoc_iso_c.f90 [hunk 10]
```diff
@@ -497,4 +497,18 @@ subroutine linear_interp_c(x1,x2,y1,y2,km1,km2,ncol,minthresh) bind (C)
 
   end subroutine linear_interp_c  
 
+ subroutine shoc_diag_second_moments_lbycond_c(shcol, wthl, wqw, uw, vw,ustar2, wstar, &
+            wthlo, wqwo, uwo, vwo, wtkeo, thlo, qwo, qwthlo) bind(C)
+   use shoc, only: diag_second_moments_lbycond
+
+   ! argmens
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/shoc_iso_c.f90:504]: Let's just get rid of these kinds of comments.

File: components/scream/src/physics/shoc/shoc_iso_f.f90 [hunk 11]
```diff
@@ -43,6 +43,17 @@ subroutine shoc_diag_second_moments_srf_f(shcol, wthl, uw, vw, ustar2, wstar) bi
 
  end subroutine shoc_diag_second_moments_srf_f
 
+ subroutine shoc_diag_second_moments_lbycond_f(shcol, wthl, wqw, uw, vw, ustar2, wstar, &
+            wthlo, wqwo, uwo, vwo, wtkeo, thlo, qwo, qwthlo) bind(C)
+   use iso_c_binding
+
+   ! argments
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/shoc_iso_f.f90:50]: Omit.

File: components/scream/src/physics/shoc/tests/shoc_diag_second_mom_lbycond_test.cpp (hunk 8 above)
Comment (jeff-cohere) [reply components/scream/src/physics/shoc/tests/shoc_diag_second_mom_lbycond_test.cpp:37]: This also applies to the test in PR #532 when we introduce non-zero tendencies.

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp [hunk 12]
```diff
@@ -495,5 +507,69 @@ void shoc_diag_second_moments_srf_f(Int shcol, Real* wthl, Real* uw, Real* vw, R
...
+  view_1d wthlo_d  ("wthl", shcol),
+          wqwo_d   ("wqw" , shcol),
+          uwo_d    ("uw"  , shcol),
+          vwo_d    ("vw"  , shcol),
+          wtkeo_d  ("wtke", shcol),
+          thlo_d   ("thl" , shcol),
+          qwo_d    ("qw"  , shcol),
+          qwthlo_d ("qwthl", shcol);
+
+  Kokkos::parallel_for("parallel_moments_lbycond", shcol, KOKKOS_LAMBDA (const int& i) {
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/shoc_functions_f90.cpp:539]: Curious what folks think about using `k` as the vertical index by convention, since we often talk about `i` as being the column index. Just a thought.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp (hunk 2 above)
Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp]: yes, done!

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp (hunk 3 above)
Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp]: cube and square function doesn't support Scalar type

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp (hunk 5 above)
Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp]: change to use sp function. done!

File: components/cam/src/physics/cam/shoc.F90 (hunk 6 above)
Comment (xyuan) [reply components/cam/src/physics/cam/shoc.F90]: square and cube function doesn't support Scalar type

File: components/scream/src/physics/shoc/tests/shoc_diag_second_mom_lbycond_test.cpp (hunk 8 above)
Comment (xyuan) [reply components/scream/src/physics/shoc/tests/shoc_diag_second_mom_lbycond_test.cpp:37]: yes, add the call to randomize methods

File: components/scream/src/physics/shoc/shoc_iso_c.f90 (hunk 10 above)
Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_iso_c.f90]: done

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp (hunk 12 above)
Comment (jgfouca) [reply components/scream/src/physics/shoc/shoc_functions_f90.cpp:539]: @jeff-cohere , he's iterating over schol (across columns), so I think `i` makes sense here.

**Comment by xyuan** [comment]
please review this PR again, all the comments are resolved now.

**Comment by xyuan** [comment]
I am closing this PR since this PR is redone.
